### PR TITLE
osd/scrub: fix the handling of interval changes

### PR DESCRIPTION
**Pull request by ronen-fr**
Separating and clarifying the handling of interval termination, new interval starting, and configuration changes:

A primary PG now registers with its OSD for scrubbing only when fully activated: on_pg_activate() called from PG::on_activate();

When the interval ends, the PG is removed from the OSD queue by stop_active_scrubs(), called from start_peering_interval().

Configuration changes are still handled by update_scrub_job().

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Review by athanatos — CHANGES_REQUESTED** [review]
Seems reasonable structurally, see my comments.

This PR's single commit seems to contain several orthogonal changes.  In order to review it, I needed to split it into smaller atomic commits.  https://github.com/athanatos/ceph/tree/sjust/wip-scrub-interval-pr-49959 has those commits -- please split up this PR in a similar way.  In the future, it would really expedite review if you'd split up your PRs in a similar manner.

File: src/osd/PeeringState.h [hunk 1]
```diff
@@ -282,15 +282,16 @@ class PeeringState : public MissingLoc::MappingInfo {
...
-
     /// Need to reschedule next scrub. Assuming no change in role
     virtual void reschedule_scrub() = 0;
 
     /// Notify that a scrub has been requested
     virtual void scrub_requested(scrub_level_t scrub_level, scrub_type_t scrub_type) = 0;
 
+    /// stop any active scrubbing (on interval end) and unregister from
+    /// the OSD scrub queue
+    virtual void stop_active_scrubs() = 0;
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/PeeringState.h:293]: I fully accept your 1'st comment. Moved stop_active_scrubs() call to PG::on_new_interval().
Re both comments: the intention here is to separate the handling of a change of interval into two parts: stopping what the scrubber was doing in the prev interval as one step, executed early on the interval transition, and a separate handling of "we have a new primary with its updated actset, which should be queued for scrubbing" - which should only be called when the PG is really ready.

The old code handled both at the same time (and the same function), which:
- caused multiple aborted attempts to "let's start scrubb'in" during new interval peering (had to be restarted, for example, each time a replica was acked and joined), and
- was confusing, and
- was even more confusing when I mixed it with 'history/conf' changes, that should only affect the scheduling of the next scrub.

That is why I maintain the separate names for the separate stages.
(and why I originally placed a direct call to stop_active_scrubs() in the peering-state code: I wanted to guarantee that that code is always executed early in the process.)

File: src/osd/scrubber/pg_scrubber.cc [hunk 2]
```diff
@@ -2599,8 +2640,6 @@ void ReplicaReservations::send_all_done()
 
 void ReplicaReservations::send_reject()
 {
-  // stop any pending timeout timer
-  m_no_reply.reset();
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:2603]: Not really - it was copied from a separate PR. But it's hard to test this PR without it, as the OSDs crash.
Will move it to a separate commit, to be rmed once that other PR is merged.

File: src/osd/scrubber/pg_scrubber.cc [hunk 3]
```diff
@@ -2374,15 +2407,22 @@ void PgScrubber::reset_internal_state()
...
+
   m_current_token++;
 
   // when advance_token() is called, it is assumed that no scrubbing takes
   // place. We will, though, verify that. And if we are actually still handling
   // a stale request - both our internal state and the FSM state will be
   // cleared.
-  replica_handling_done();
-  m_fsm->process_event(FullReset{});
+  if (m_fsm->is_active_replica()) {
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:2421]: the full_reset event is always effective, regardless of the FSM state (as it was meant
as a way to guarantee reset). Note the next line, which I do not want to call unless
active as replica. I could have used a new event, and moved the replica_handling_done()
to the FSM reactions, but (for me at least) the added code would have just made the logic
less clear.

File: src/osd/PeeringState.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (athanatos) [reply src/osd/PeeringState.h]: Ah, I was unclear.   I'd expect to see two scrubber side methods: one for when an interval ends and another for when the pg goes active.  I think you've already got the one for when the pg goes active, though.

File: src/osd/scrubber/pg_scrubber.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:2603]: Not directly - but I was unable to test the PR without this fix.
Moved into a separate commit.

File: src/osd/scrubber/pg_scrubber.cc [hunk 4]
```diff
@@ -488,69 +531,56 @@ std::string_view PgScrubber::registration_state() const
...
+	   << dendl;
 }
 
 void PgScrubber::update_scrub_job(const requested_scrub_t& request_flags)
 {
   dout(10) << fmt::format("{}: flags:<{}>", __func__, request_flags) << dendl;
   // verify that the 'in_q' status matches our "Primariority"
   if (m_scrub_job && is_primary() && !m_scrub_job->in_queues) {
-    dout(1) << __func__ << " !!! primary but not scheduled! " << dendl;
+    dout(10) << __func__ << " !!! primary but not scheduled! " << dendl;
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc]: For now - left this change out of the PR

**Comment by ronen-fr** [comment]
@athanatos thanks for the work invested, especially re the commit splits

File: src/osd/scrubber/pg_scrubber.cc [hunk 5]
```diff
@@ -2599,8 +2634,6 @@ void ReplicaReservations::send_all_done()
 
 void ReplicaReservations::send_reject()
 {
-  // stop any pending timeout timer
-  m_no_reply.reset();
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:2603]: @athanatos : yes - that was the failed scenario. Checking what you said in the last sentence.

File: src/osd/scrubber/pg_scrubber.cc [hunk 6]
```diff
@@ -488,53 +531,41 @@ std::string_view PgScrubber::registration_state() const
...
-  }
 
-  // is there an interval change we should respond to?
-  if (is_primary() && is_scrub_active()) {
-    if (m_interval_start < m_pg->get_same_interval_since()) {
-      dout(10) << fmt::format(
-		    "{}: interval changed ({} -> {}). Aborting active scrub.",
-		    __func__, m_interval_start, m_pg->get_same_interval_since())
-	       << dendl;
-      scrub_clear_state();
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:527]: By "a message carrying the request to be reserved by a primary was queued", do you mean the time the message spends between being queued via the messenger Dispatcher interface and when it gets processed by the PG and and scrubber?  Why is that important?

File: src/osd/scrubber/pg_scrubber.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:2603]: I'm now less sure about this diagnosis.  SafeTimer can optionally either hold the lock while calling callbacks or not.  There's a boolean param passed to the constructor which governs this behavior -- safe_callbacks.  sleep_timer specifically seems to be constructed with safe_callbacks set to false, so locking OSDService::sleep_lock shouldn't be a problem.

Can you link to the tracker id?

File: src/osd/scrubber/pg_scrubber.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:527]: @ronen-fr I think it's worth mapping these resources onto the state machine.  I've started working on a branch https://github.com/athanatos/ceph/tree/sjust/wip-scrub-interval-pr-49959-2 which does that.  I'm still working on it, so it's likely got errors.

**Pull request closed without merging**